app/services/hyrax/work_uploads_handler.rb

```diff
@@ -92,47 +82,63 @@ def add(files:, file_set_params: [])
     def attach
       return true if Array.wrap(files).empty? # short circuit to avoid aquiring a lock we won't use
 
-      acquire_lock_for(work.id) do
-        event_payloads = files.each_with_object([]).with_index do |(file, arry), index|
-          arry << make_file_set_and_ingest(file, @file_set_params[index] || {})
-        end
-        @persister.save(resource: work)
-        Hyrax.publisher.publish('object.metadata.updated', object: work, user: files.first.user)
-        event_payloads.each do |payload|
-          payload.delete(:job).enqueue
-          Hyrax.publisher.publish('file.set.attached', payload)
-        end
+      event_payloads = create_file_sets(files)
+      append_file_sets_to_work(file_ids: event_payloads.map { |payload| payload[:file_set].id }, user: files.first.user)
+      event_payloads.each do |payload|
+        payload.delete(:job).enqueue
+        Hyrax.publisher.publish('file.set.attached', { file_set: payload[:file_set], user: payload[:user] })
       end
     end
 
     private
 
     ##
     # @api private
-    def make_file_set_and_ingest(file, file_set_params = {})
-      file_set = @persister.save(resource: Hyrax::FileSet.new(file_set_args(file, file_set_params)))
-      Hyrax.publisher.publish('object.deposited', object: file_set, user: file.user)
-      file.add_file_set!(file_set)
+    def create_file_sets(files)
+      files.each_with_object([]).with_index do |(file, arry), index|
+        file_set = find_or_create_file_set(file, @file_set_params[index] || {})
+        update_file_set(file_set, file)
+        arry << { file_set: file_set, user: file.user, job: ValkyrieIngestJob.new(file) }
+      end
+    end
+
+    ##
+    # @api private
+    def append_file_sets_to_work(file_ids:, user:)
```

Comment:
Can this live in the Work?

Reply:
This has to reload the @work instance variable, so I think we can't just move it part and parcel to the Work.